clients/src/main/java/org/apache/kafka/clients/consumer/OffsetAndMetadata.java

```diff
@@ -19,23 +19,39 @@
 import org.apache.kafka.common.requests.OffsetFetchResponse;
 
 import java.io.Serializable;
+import java.util.Objects;
+import java.util.Optional;
 
 /**
  * The Kafka offset commit API allows users to provide additional metadata (in the form of a string)
  * when an offset is committed. This can be useful (for example) to store information about which
  * node made the commit, what time the commit was made, etc.
  */
 public class OffsetAndMetadata implements Serializable {
+    private static final long serialVersionUID = 2019555404968089681L;
+
     private final long offset;
     private final String metadata;
 
+    // We use null to represent the absence of a leader epoch to simplify serialization.
+    // I.e., older serializations of this class which do not have this field will automatically
+    // initialize its value to null.
+    private final Integer leaderEpoch;
```

Comment:
That was what I attempted to do initially. Unfortunately we cannot depend on initializers during object deserialization. When deserializing old versions, the new leader epoch field will be initialized to the default value for the corresponding type. So if we use `int leaderEpoch = RecordBatch.NO_PARTITION_LEADER_EPOCH`, then the value after deserialization will actually be 0. Since 0 is a valid value for this field, I couldn't think of a way to make it work. On the other hand, the default value for `Integer` is null, so it works. 

I really wish we hadn't made this class `Serializable`, but tough to take it away now. One possibility is to make this an exception case and to prefer the use of the `int` type for other classes.

Reply:
@hachikuji You could probably use `readObject` if you want right?

Reply:
I don't think so. At least I could not find a way to make it work. If you override `readObject`, you still have to call `defaultReadObject`, but this will initialize all of the fields. But then how do you know whether the initialized value was read from the serialized data or is simply the default value?

Reply:
Ah I see. I agree it is simpler to leave it as is.